### PR TITLE
final project 4-th sprint

### PR DESCRIPTION
**Pull request by StepanKazakov**

### CODE REVIEW

File: conftest.py [hunk 1]
```diff
@@ -0,0 +1,20 @@
+import pytest
+from main import BooksCollector
+
+
+@pytest.fixture(scope='function')
+def collector():
+    collector = BooksCollector()
```
Comment (irisqul) [inline conftest.py:7]: Отлично: в каждом тесте создается собственный экземпляр BooksCollector

File: tests.py [hunk 2]
```diff
@@ -1,24 +1,82 @@
...
+    @pytest.mark.parametrize('name', [
+        'I',                                         # Single character, boundary value
+        'Проверка максимальной длины названия рус',  # Maximum valid length, boundary value
+        'Sherlock Holmes',                           # Normal valid name
+        '1984',                                      # Numeric characters in string format
+        "'@#$%^&*()'"                                # Special characters in string format
+    ])
+    def test_add_new_book_check_boundary_values_positive(self, collector, my_books, name):
+        collector.add_new_book(name)
+        assert len(my_books) == 1
```
Comment (irisqul) [inline tests.py:16]: Нужно исправить: фикстура - это предусловие теста, она не должна заменять шаги и мы не вызываем фикстуры после совершения шагов. Лучше будет убрать my_books my_favorite фикстуры и проста совершать запрос для получения данных как шаг теста

File: tests.py [hunk 3]
```diff
@@ -1,24 +1,82 @@
...
+
+    def test_set_book_genre_set_nonexistent_genre_to_book_result_not_set(self, collector, my_books):
+        collector.add_new_book('Sherlock Holmes')
+        collector.set_book_genre('Sherlock Holmes', 'Детектив')
+        assert my_books['Sherlock Holmes'] == ''
+
+    def test_set_book_genre_change_genre_return_last_value(self, collector, my_books):
+        collector.add_new_book('1984')
+        collector.set_book_genre('1984', 'Фантастика')
+        assert my_books['1984'] == 'Фантастика'
```
Comment (irisqul) [inline tests.py]: Нужно исправить: тест неатомарный. после ассерта уже не должно быть шагов, а каждый тест должен проверять какое то одно поведение приложения

File: tests.py [hunk 4]
```diff
@@ -1,24 +1,82 @@
...
+        assert 'Metro 2033' not in children_books
 
-        # проверяем, что добавилось именно две
-        # словарь books_rating, который нам возвращает метод get_books_rating, имеет длину 2
-        assert len(collector.get_books_rating()) == 2
+    def test_add_book_in_favorites_check_and_remove(self, collector, my_favorite):
+        assert my_favorite == []
+        collector.add_new_book('Metro 2035')
+        collector.add_book_in_favorites('Metro 2035')
+        assert 'Metro 2035' in my_favorite
```
Comment (irisqul) [inline tests.py]: up

File: tests.py [hunk 5]
```diff
@@ -1,24 +1,82 @@
...
+        collector.delete_book_from_favorites('Metro 2035')
+        assert 'Metro 2035' not in my_favorite
 
-    # напиши свои тесты ниже
-    # чтобы тесты были независимыми в каждом из них создавай отдельный экземпляр класса BooksCollector()
\ No newline at end of file
+    def test_add_book_in_favorites_double_add_list_of_favorites_return_one(self, collector, my_favorite):
+        assert my_favorite == []
```
Comment (irisqul) [inline tests.py]: Нужно исправить: мы тестируем не атрибуты экземпляров класса, а методы приложения. Необходимо добавить шаги в которых будут использованы методы и проверить результат.
